js-pkg/client/src/provider.ts

```diff
@@ -143,21 +154,83 @@ export class YSweetProvider {
     this.awareness.on('update', this.handleAwarenessUpdate.bind(this))
     this.WebSocketPolyfill = extraOptions.WebSocketPolyfill || WebSocket
 
+    this.online = this.online.bind(this)
+    this.offline = this.offline.bind(this)
+    if (typeof window !== 'undefined') {
+      window.addEventListener('offline', this.offline)
+      window.addEventListener('online', this.online)
+    }
+
     doc.on('update', this.update.bind(this))
 
     if (extraOptions.connect !== false) {
       this.connect()
     }
   }
 
+  private offline() {
+    // When the browser indicates that we are offline, we immediately
+    // probe the connection status.
+    // This accelerates the process of discovering we are offline, but
+    // doesn't mean we entirely true the browser, since it can be wrong
+    // (e.g. in the case that the connection is over localhost).
+    this.checkSync()
+    this.setConnectionTimeout()
+  }
+
+  private online() {
+    if (this.reconnectSleeper) {
+      this.reconnectSleeper.wake()
+    }
+  }
+
+  private clearHeartbeat() {
+    if (this.heartbeatHandle) {
+      clearTimeout(this.heartbeatHandle)
+      this.heartbeatHandle = null
+    }
+  }
+
+  private setHeartbeat() {
+    if (this.heartbeatHandle) {
+      return
+    }
+    this.heartbeatHandle = setTimeout(() => {
+      this.checkSync()
+      this.setConnectionTimeout()
+      this.heartbeatHandle = null
+    }, MAX_TIMEOUT_BETWEEN_HEARTBEATS)
+  }
```

Comment:
Trying to think if there's ever a scenario in which we want to bail out of setting a heartbeat if there's one queued, rather than just resetting it? I think we basically always want a heartbeat going and to just keep "moving the timer up" when we receive messages from the server.

I'd change these methods like this:

```suggestion
  private clearHeartbeat() {
    if (this.heartbeatHandle) {
      clearTimeout(this.heartbeatHandle)
      this.heartbeatHandle = null
    }
  }

  private resetHeartbeat() {
    this.clearHeartbeat()

    this.heartbeatHandle = setTimeout(() => {
      this.checkSync()
      this.setConnectionTimeout()
      this.heartbeatHandle = null
    }, MAX_TIMEOUT_BETWEEN_HEARTBEATS)
  }
```

And then we should probably call `clearHeartbeat` in `websocketClose` and `websocketError`.